application/frontend/controllers/AuthController.php

```diff
@@ -0,0 +1,123 @@
+    }
+
+    public function actionLogin()
+    {
+        /*
+         * Collect return to url of where to send user after successful login
+         * Expected as relative url starting with /
+         * Before redirecting user after login this will be prefixed with ui_url
+         */
+        $returnTo = \Yii::$app->request->get('ReturnTo', '');
+
+        if ( ! \Yii::$app->user->isGuest) {
+            $afterLogin = $this->getAfterLoginUrl($returnTo);
+            return $this->redirect($afterLogin);
+        }
+
+        /*
+         * Initialize $log variable for logging
+         */
+        $log = ['action' => 'login'];
+
+        try {
+            $returnTo = Url::to(['auth/login', 'ReturnTo' => $returnTo], true);
+            /** @var AuthUser $authUser */
+            $authUser = \Yii::$app->auth->login(\Yii::$app->request, $returnTo);
+
+            $log['email'] = $authUser->email;
+            // Get local user instance or create one
+            $user = User::findOrCreate(null, null, $authUser->employeeId);
+            // Initialize session for user
+            if (\Yii::$app->user->login($user, \Yii::$app->params['sessionDuration'])) {
+                $log['status'] = 'success';
+                \Yii::warning($log, 'application');
+
+                $afterLogin = $this->getAfterLoginUrl($returnTo);
+                return $this->redirect($afterLogin);
+            } else {
+                throw new UnauthorizedHttpException('Unable to perform user login', 1459966846);
+            }
+        } catch (RedirectException $e) {
+            /*
+             * Login triggered redirect to IdP to login, so return a redirect to it
+             */
+            return $this->redirect($e->getUrl());
+        } catch (\Exception $e) {
+            $log['status'] = 'error';
+            $log['error'] = $e->getMessage();
+            $log['code'] = $e->getCode();
+            \Yii::error($log, 'application');
+            throw new UnauthorizedHttpException($e->getMessage(), $e->getCode());
+        }
+
+    }
+
+    public function actionLogout()
+    {
+        if ( ! \Yii::$app->user->isGuest) {
+            $authUser = \Yii::$app->user->identity->getAuthUser();
+        } else {
+            /*
+             * User not logged in, but lets kill session anyway and redirect to UI
+             */
+            \Yii::$app->user->logout(true);
+            return $this->redirect(\Yii::$app->params['ui_url']);
+        }
+
+        /*
+         * Kill local session
+         */
+        \Yii::$app->user->logout(true);
+
+        /*
+         * Log user out of IdP
+         */
+        try {
+            \Yii::$app->auth->logout($authUser, \Yii::$app->params['ui_url']);
+        } catch (RedirectException $e) {
+            return $this->redirect($e->getUrl());
+        }
+
+    }
+
+    public function getAfterLoginUrl($returnTo)
+    {
+        /*
+         * Only keep $returnTo if it is a path on the frontend
```

Comment:
Do you think it would be worth adding something to this comment to remind us (down the road) why we only allow redirecting to relative paths?  (i.e. - to prevent phishing attacks using the actual login process to send someone to a bad/fake website, etc.)